### PR TITLE
convert `expr_walk*` to use iteration instead of recursion

### PR DESCRIPTION
**Pull request by pedrocarlo**
## Description
Depends on #6636 
Changes the `expr_walk*` functions to be iterative instead. There is some unsafe around the `walk_expr_mut`, but in this case here we are just copying the same traversal order and logic as in the recursive case, so it should be safe.

## Motivation and context
Reduce stack usage and related to the change I want to make here to avoid stack overflows: https://github.com/tursodatabase/turso/pull/6634

## Description of AI Usage
Mechanical Pass with Codex. It was surprisingly simple to do.

### CODE REVIEW

**Comment by LeMikaelF** [comment]
Could you add a test with a deep AST, to guard against stack overflow regressions?

**Comment by LeMikaelF** [comment]
Do we have benchmarks for this? I'm a bit skeptical of the 1-sized `SmallVec`, any `Binary`, `Between`, or other nodes that embed more than one `Expr` will use the heap.

**Comment by LeMikaelF** [comment]
My Claude Code rewrote this without `unsafe`, I didn't look at it in detail but it says that the performance is equivalent. We should probably avoid `unsafe` if we can.

Here's the patch:

```
Index: core/translate/expr/walk.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/translate/expr/walk.rs b/core/translate/expr/walk.rs
--- a/core/translate/expr/walk.rs	(revision 38b8e4207fc7453da54ceae67a7a0eb39738b3d8)
+++ b/core/translate/expr/walk.rs	(date 1778010207298)
@@ -210,69 +210,60 @@
 }
 
 /// Walks a mutable expression, applying a function to each sub-expression.
-pub fn walk_expr_mut<F>(expr: &mut ast::Expr, func: &mut F) -> Result<WalkControl>
+pub fn walk_expr_mut<'a, F>(expr: &'a mut ast::Expr, func: &mut F) -> Result<WalkControl>
 where
     F: FnMut(&mut ast::Expr) -> Result<WalkControl>,
 {
-    enum WalkMutItem {
-        Expr(*mut ast::Expr),
-        FrameBound(*mut ast::FrameBound),
-    }
-
-    type Stack = smallvec::SmallVec<[WalkMutItem; 1]>;
-
-    fn expr_item(expr: &mut ast::Expr) -> WalkMutItem {
-        WalkMutItem::Expr(expr as *mut ast::Expr)
+    enum WalkItem<'a> {
+        Expr(&'a mut ast::Expr),
+        FrameBound(&'a mut ast::FrameBound),
     }
 
-    fn frame_bound_item(bound: &mut ast::FrameBound) -> WalkMutItem {
-        WalkMutItem::FrameBound(bound as *mut ast::FrameBound)
-    }
+    type Stack<'a> = smallvec::SmallVec<[WalkItem<'a>; 1]>;
 
-    fn push_over_clause_walk_items(stack: &mut Stack, over_clause: &mut ast::Over) {
+    fn push_over_clause_walk_items<'a>(
+        stack: &mut Stack<'a>,
+        over_clause: &'a mut ast::Over,
+    ) {
         if let ast::Over::Window(window) = over_clause {
             if let Some(frame_clause) = &mut window.frame_clause {
                 if let Some(end_bound) = &mut frame_clause.end {
-                    stack.push(frame_bound_item(end_bound));
+                    stack.push(WalkItem::FrameBound(end_bound));
                 }
-                stack.push(frame_bound_item(&mut frame_clause.start));
+                stack.push(WalkItem::FrameBound(&mut frame_clause.start));
             }
             for sort_col in window.order_by.iter_mut().rev() {
-                stack.push(expr_item(&mut sort_col.expr));
+                stack.push(WalkItem::Expr(&mut sort_col.expr));
             }
             for part_expr in window.partition_by.iter_mut().rev() {
-                stack.push(expr_item(part_expr));
+                stack.push(WalkItem::Expr(part_expr));
             }
         }
     }
 
-    let mut stack: Stack = smallvec::smallvec![expr_item(expr)];
+    let mut stack: Stack<'a> = smallvec::smallvec![WalkItem::Expr(expr)];
     while let Some(item) = stack.pop() {
         match item {
-            WalkMutItem::Expr(expr) => {
-                // SAFETY: The stack only stores pointers to child fields of nodes
-                // that have already been popped, so no two live `&mut Expr`
-                // aliases are created.
-                let expr = unsafe { &mut *expr };
+            WalkItem::Expr(expr) => {
                 if matches!(func(expr)?, WalkControl::SkipChildren) {
                     continue;
                 }
                 match expr {
                     ast::Expr::SubqueryResult { lhs, .. } => {
                         if let Some(lhs) = lhs {
-                            stack.push(expr_item(lhs));
+                            stack.push(WalkItem::Expr(lhs));
                         }
                     }
                     ast::Expr::Between {
                         lhs, start, end, ..
                     } => {
-                        stack.push(expr_item(end));
-                        stack.push(expr_item(start));
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(end));
+                        stack.push(WalkItem::Expr(start));
+                        stack.push(WalkItem::Expr(lhs));
                     }
                     ast::Expr::Binary(lhs, _, rhs) => {
-                        stack.push(expr_item(rhs));
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(rhs));
+                        stack.push(WalkItem::Expr(lhs));
                     }
                     ast::Expr::Case {
                         base,
@@ -280,18 +271,18 @@
                         else_expr,
                     } => {
                         if let Some(else_expr) = else_expr {
-                            stack.push(expr_item(else_expr));
+                            stack.push(WalkItem::Expr(else_expr));
                         }
                         for (when_expr, then_expr) in when_then_pairs.iter_mut().rev() {
-                            stack.push(expr_item(then_expr));
-                            stack.push(expr_item(when_expr));
+                            stack.push(WalkItem::Expr(then_expr));
+                            stack.push(WalkItem::Expr(when_expr));
                         }
                         if let Some(base_expr) = base {
-                            stack.push(expr_item(base_expr));
+                            stack.push(WalkItem::Expr(base_expr));
                         }
                     }
                     ast::Expr::Cast { expr, .. } | ast::Expr::Collate(expr, _) => {
-                        stack.push(expr_item(expr));
+                        stack.push(WalkItem::Expr(expr));
                     }
                     ast::Expr::Exists(_) | ast::Expr::Subquery(_) => {
                         // TODO: Walk through select statements if needed
@@ -306,13 +297,13 @@
                             push_over_clause_walk_items(&mut stack, over_clause);
                         }
                         if let Some(filter_clause) = &mut filter_over.filter_clause {
-                            stack.push(expr_item(filter_clause));
+                            stack.push(WalkItem::Expr(filter_clause));
                         }
                         for sort_col in order_by.iter_mut().rev() {
-                            stack.push(expr_item(&mut sort_col.expr));
+                            stack.push(WalkItem::Expr(&mut sort_col.expr));
                         }
                         for arg in args.iter_mut().rev() {
-                            stack.push(expr_item(arg));
+                            stack.push(WalkItem::Expr(arg));
                         }
                     }
                     ast::Expr::FunctionCallStar { filter_over, .. } => {
@@ -320,49 +311,49 @@
                             push_over_clause_walk_items(&mut stack, over_clause);
                         }
                         if let Some(filter_clause) = &mut filter_over.filter_clause {
-                            stack.push(expr_item(filter_clause));
+                            stack.push(WalkItem::Expr(filter_clause));
                         }
                     }
                     ast::Expr::InList { lhs, rhs, .. } => {
                         for expr in rhs.iter_mut().rev() {
-                            stack.push(expr_item(expr));
+                            stack.push(WalkItem::Expr(expr));
                         }
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(lhs));
                     }
                     ast::Expr::InSelect { lhs, rhs: _, .. } => {
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(lhs));
                         // TODO: Walk through select statements if needed
                     }
                     ast::Expr::InTable { lhs, args, .. } => {
                         for expr in args.iter_mut().rev() {
-                            stack.push(expr_item(expr));
+                            stack.push(WalkItem::Expr(expr));
                         }
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(lhs));
                     }
                     ast::Expr::IsNull(expr) | ast::Expr::NotNull(expr) => {
-                        stack.push(expr_item(expr));
+                        stack.push(WalkItem::Expr(expr));
                     }
                     ast::Expr::Like {
                         lhs, rhs, escape, ..
                     } => {
                         if let Some(esc_expr) = escape {
-                            stack.push(expr_item(esc_expr));
+                            stack.push(WalkItem::Expr(esc_expr));
                         }
-                        stack.push(expr_item(rhs));
-                        stack.push(expr_item(lhs));
+                        stack.push(WalkItem::Expr(rhs));
+                        stack.push(WalkItem::Expr(lhs));
                     }
                     ast::Expr::Parenthesized(exprs) => {
                         for expr in exprs.iter_mut().rev() {
-                            stack.push(expr_item(expr));
+                            stack.push(WalkItem::Expr(expr));
                         }
                     }
                     ast::Expr::Raise(_, expr) => {
                         if let Some(raise_expr) = expr {
-                            stack.push(expr_item(raise_expr));
+                            stack.push(WalkItem::Expr(raise_expr));
                         }
                     }
                     ast::Expr::Unary(_, expr) => {
-                        stack.push(expr_item(expr));
+                        stack.push(WalkItem::Expr(expr));
                     }
                     ast::Expr::Array { .. } | ast::Expr::Subscript { .. } => {
                         unreachable!(
@@ -380,23 +371,18 @@
                     | ast::Expr::Register(_)
                     | ast::Expr::Default => {}
                     ast::Expr::FieldAccess { base, .. } => {
-                        stack.push(expr_item(base));
+                        stack.push(WalkItem::Expr(base));
                     }
                 }
             }
-            WalkMutItem::FrameBound(bound) => {
-                // SAFETY: Frame-bound pointers are pushed from the same popped
-                // parent node discipline as expression pointers above.
-                let bound = unsafe { &mut *bound };
-                match bound {
-                    ast::FrameBound::Following(expr) | ast::FrameBound::Preceding(expr) => {
-                        stack.push(expr_item(expr));
-                    }
-                    ast::FrameBound::CurrentRow
-                    | ast::FrameBound::UnboundedFollowing
-                    | ast::FrameBound::UnboundedPreceding => {}
-                }
-            }
+            WalkItem::FrameBound(bound) => match bound {
+                ast::FrameBound::Following(expr) | ast::FrameBound::Preceding(expr) => {
+                    stack.push(WalkItem::Expr(expr));
+                }
+                ast::FrameBound::CurrentRow
+                | ast::FrameBound::UnboundedFollowing
+                | ast::FrameBound::UnboundedPreceding => {}
+            },
         }
     }
     Ok(WalkControl::Continue)
```

**Review by LeMikaelF — APPROVED** [review]
